readme.md
# navigation.js

-   Install: `npm install navigation.js --save`
-   Test: `npm test`
-   Run debug map locally: `npm start`. Then head to `http://localhost:9966`

Example:

```js
var navigation = require('navigation.js')({
    units: 'miles',
    maxReRouteDistance: 0.03,
    maxSnapToLocation: 0.01
});

// Where is the user along the route?
navigation.getCurrentStep(userLocation, mapboxDirectionRoute.routes[0].legs[0], 0);
```

## API Usage

### getCurrentStep

Given a user location and route, calculates closest step to user.

**Parameters**

-   `user` **object** point feature representing user location. Must be a valid GeoJSON object.
-   `route` **object** from [Mapbox directions API](https://www.mapbox.com/developers/api/directions/).
    The Mapbox directions API returns an object with up to 2 `routes` on the `route` key. `getCurrentStep` expects of these routes, either the first or second.
-   `userCurrentStep` **number** along the route
-   `userBearing` **number** current user bearing. If provided, the user must be within a certain threshold of the steps exit bearing to successful complete a step.

Returns **object** Containing 3 keys: `step`, `distance`, `snapToLocation`. `distance` is the line distance to end of step, `absoluteDistance` is the users absolute distance to the end of the route `snapToLocation` is location along route which is closest to the user.

### options

Configuration options

**Parameters**

-   `object`  `units` - either `miles` or `km`. `maxReRouteDistance` - max distance the user can be from the route. `maxSnapToLocation` - max distance to snap user to route. `completionDistance` - distance away from end of step that is considered a completion. If this distance is shorter than the step distance, it will be changed to 10ft. `warnUserTime` - number of seconds ahead of maneuver to warn user about maneuver. `shortCompletionDistance` - if the step is shorter than the `completionDistance`, this distance will be used to calculate if the step has been completed. `userBearingCompleteThreshold` - Bearing threshold for the user to complete a step


CHANGELOG.md
# v1.1.0

* Mapbox Directions API v5 is now required
* Adds support for `geometries=polyline`
* Changes `findNextStep` -> `getCurrentStep`

# v1.0.0

* Removes `shouldReRoute` method. It can now be found as a key on `findNextStep`.
* Added 3rd required argument to `findNextStep` which is the users current step.
* Internal logic has changed about whether the user should be rerouted. Now, the user must be within x miles of the given step, not the entire route.
* Added new key `alertUserLevel` which is an object with keys, `high` and `low`. `high` and `low` both return booleans which are false until the user should be notified about an upcoming maneuver.

# v0.2.0

* `distance` is no line distance.
* Adds `absoluteDistance` which is the users absolute distance to the end of the step.

# v0.1.0

* `maxDistance` changed to `maxReRouteDistance`.
* Added option `maxSnapToLocation`.
* `findNextStep` returns snapped location to route now on key `snapToLocation`.

# v0.0.1

* Initial release


tests/debug/index.js
require('mapbox.js');
require('leaflet-hash');
var request = require('request');
var debounce = require('debounce');
var endpoint = 'https://api.mapbox.com/directions/v5/mapbox/driving/';
var polyline = require('polyline');

// Setup map
L.mapbox.accessToken = 'pk.eyJ1IjoiYm9iYnlzdWQiLCJhIjoiTi16MElIUSJ9.Clrqck--7WmHeqqvtFdYig';
var map = L.mapbox.map('map').setView([39.9229, -75.1351], 14);
L.mapbox.styleLayer('mapbox://styles/mapbox/streets-v8').addTo(map);
L.hash(map);
var marker = L.marker([0, 0]).addTo(map);
var routeGeoJSON = L.layerGroup().addTo(map);
var userLocation = { type: 'Feature', properties: {}, geometry: { type: 'Point', coordinates: [] }};
L.geoJson(require('../fixtures/downtown').routes[0].geometry).addTo(routeGeoJSON);

var activeRoute = require('../fixtures/downtown').routes[0];
var stepDiv = document.getElementById('step');

// Navigation.js can handle both geometries=polyline and geometries=geojson
// However, the Mapbox directions API will default to geometries=geojson
var geometries = 'polyline';
// var geometries = 'geojson';

// Setup navigation.js
var navigation = require('../../')({
    units: 'miles',
    maxReRouteDistance: 0.03,
    maxSnapToLocation: 0.01
});

// Alwats set the initial step to 0
var currentStep = 0;

map.on('mousemove', function(e) {
    userLocation.geometry.coordinates[0] = e.latlng.lng;
    userLocation.geometry.coordinates[1] = e.latlng.lat;

    var stepInfo = navigation.getCurrentStep(userLocation, activeRoute.legs[0], currentStep);

    // Display whether user should reroute in UI
    document.getElementById('reroute').innerHTML = stepInfo.shouldReRoute;

    // If the user is off the route, `shouldReRoute` will be true
    // In this case, recalculate the route, clear the current route, add the new route
    // and most importantly, reset the currentStep to 0
    if (stepInfo.shouldReRoute) {
        getRoute(e.latlng.lng, e.latlng.lat, -75.118674, 39.94156, function(err, route) {
            routeGeoJSON.clearLayers();
            stepDiv.classList.remove('flash');

            // This example shows how to handle both encoded polylines and GeoJSON
            if (typeof route.routes[0].geometry === 'string') {
                L.geoJson(polyline.toGeoJSON(route.routes[0].geometry)).addTo(routeGeoJSON);
            } else {
                L.geoJson(route.routes[0].geometry).addTo(routeGeoJSON);
            }

            activeRoute = route.routes[0];
            currentStep = 0;
        });
    }

    // Flash the instructions to signal to the user the maneuver is coming up soon
    if (stepInfo.alertUserLevel.high) {
        if (!stepDiv.classList.contains('flash')) stepDiv.classList.add('flash');
    }

    // If the calculated step is greater than the users current step, update it
    // This means the user completed the current step. Also, turn off flashing
    if (stepInfo.step > currentStep) {
        currentStep = stepInfo.step;
        stepDiv.classList.remove('flash');
    }

    // Get the instruction of the next step and display it
    document.getElementById('step').innerHTML = 'In ' + Math.round(stepInfo.distance * 5280) + ' ft '+ activeRoute.legs[0].steps[stepInfo.step + 1].maneuver.instruction;

    // Snap the marker to closest point along the route
    marker.setLatLng([stepInfo.snapToLocation.geometry.coordinates[1], stepInfo.snapToLocation.geometry.coordinates[0]]);
});

var getRoute = debounce(function(fromLng, fromLat, toLng, toLat, callback) {
    request(endpoint + fromLng + ',' + fromLat + ';' + toLng + ',' + toLat + '.json?geometries=' + geometries + '&overview=full&steps=true&access_token=pk.eyJ1IjoiYm9iYnlzdWQiLCJhIjoiTi16MElIUSJ9.Clrqck--7WmHeqqvtFdYig', function(err, res, body) {
        if (err) return callback(err);
        if (body && res.statusCode === 200) return callback(null, JSON.parse(body));
    });
}, 1000);


tests/nextStep.test.js
var test = require('tape');
var route = require('./fixtures/route');

test('userToRoute should reRoute', function(t) {
    var navigation = require('../').nextStep({ units: 'miles' });
    var user = {'type': 'Feature', 'geometry': {'type': 'Point', 'coordinates': [-75.22331008911133, 39.94409450436698]}};
    var nav = navigation.findNextStep(user, route.routes[0].legs[0], 0);
    t.equal(nav.shouldReRoute, true);
    t.equal(nav.step, 0);
    t.equal(nav.absoluteDistance, 5.391214204745432);
    t.equal(nav.distance, 0.09073297470143911);
    t.equal(nav.alertUserLevel.low, false);
    t.equal(nav.alertUserLevel.high, false);
    t.equal(nav.snapToLocation.geometry.coordinates[0], -75.22331008911134);
    t.equal(nav.snapToLocation.geometry.coordinates[1], 39.94409450436698);
    t.end();
});

test('userToRoute should not reRoute', function(t) {
    var navigation = require('../').nextStep({ units: 'miles' });
    var user = {'type': 'Feature', 'geometry': {'type': 'Point', 'coordinates': [-75.12331008911133, 39.94409450436698]}};
    var nav = navigation.findNextStep(user, route.routes[0].legs[0], 0);
    t.equal(nav.shouldReRoute, false);
    t.equal(nav.step, 0);
    t.equal(nav.absoluteDistance, 0.09237455025286956);
    t.equal(nav.distance, 0.09073297470143911);
    t.equal(nav.alertUserLevel.low, false);
    t.equal(nav.alertUserLevel.high, false);
    t.equal(nav.snapToLocation.geometry.coordinates[0], -75.12331008911133);
    t.equal(nav.snapToLocation.geometry.coordinates[1], 39.94409450436698);
    t.end();
});

test('should signal next step', function(t) {
    var navigation = require('../').nextStep({ units: 'miles' });
    var user = {'type': 'Feature', 'geometry': {'type': 'Point', 'coordinates': [-75.121567, 39.944115]}};
    var nav = navigation.findNextStep(user, route.routes[0].legs[0], 0);
    t.equal(nav.shouldReRoute, false);
    t.equal(nav.step, 1);
    t.equal(nav.absoluteDistance, 0);
    t.equal(nav.distance, 0);
    t.equal(nav.alertUserLevel.low, false);
    t.equal(nav.alertUserLevel.high, true);
    t.equal(nav.snapToLocation.geometry.coordinates[0], -75.121567);
    t.equal(nav.snapToLocation.geometry.coordinates[1], 39.944115);
    t.end();
});

test('should not signal next step if bearing is provided and not within threshold', function(t) {
    var navigation = require('../').nextStep({ units: 'miles' });
    var user = {'type': 'Feature', 'geometry': {'type': 'Point', 'coordinates': [-75.121567, 39.944115]}};
    var nav = navigation.findNextStep(user, route.routes[0].legs[0], 0, 90);
    t.equal(nav.shouldReRoute, false);
    t.equal(nav.step, 0, 'step should remain zero');
    t.equal(nav.absoluteDistance, 0);
    t.equal(nav.distance, 0);
    t.equal(nav.alertUserLevel.low, false);
    t.equal(nav.alertUserLevel.high, true);
    t.equal(nav.snapToLocation.geometry.coordinates[0], -75.121567);
    t.equal(nav.snapToLocation.geometry.coordinates[1], 39.944115);
    t.end();
});

test('should signal next step if bearing is provided and is within threshold', function(t) {
    var navigation = require('../').nextStep({ units: 'miles' });
    var user = {'type': 'Feature', 'geometry': {'type': 'Point', 'coordinates': [-75.121567, 39.944115]}};
    var nav = navigation.findNextStep(user, route.routes[0].legs[0], 0, 160);
    t.equal(nav.shouldReRoute, false);
    t.equal(nav.step, 1, 'step should increment');
    t.equal(nav.absoluteDistance, 0);
    t.equal(nav.distance, 0);
    t.equal(nav.alertUserLevel.low, false);
    t.equal(nav.alertUserLevel.high, true);
    t.equal(nav.snapToLocation.geometry.coordinates[0], -75.121567);
    t.equal(nav.snapToLocation.geometry.coordinates[1], 39.944115);
    t.end();
});

test('userToRoute should not reRoute and uses kilometers', function(t) {
    var navigation = require('../').nextStep({ units: 'kilometers' });
    var user = {'type': 'Feature', 'geometry': {'type': 'Point', 'coordinates': [-75.12331008911133, 39.94409450436698]}};
    var nav = navigation.findNextStep(user, route.routes[0].legs[0], 0);
    t.equal(nav.shouldReRoute, false);
    t.equal(nav.step, 0);
    t.equal(nav.absoluteDistance, 0.14866237594988327);
    t.equal(nav.distance, 0.14602051711420994);
    t.equal(nav.alertUserLevel.low, false);
    t.equal(nav.alertUserLevel.high, false);
    t.equal(nav.snapToLocation.geometry.coordinates[0], -75.12331008911133);
    t.equal(nav.snapToLocation.geometry.coordinates[1], 39.94409450436698);
    t.end();
});


lib/nextStep.js
var turfPointOnLine = require('turf-point-on-line');
var turfDistance = require('turf-distance');
var turfLineDistance = require('turf-line-distance');
var turfLineSlice = require('turf-line-slice');
var feetToMiles = 0.000189394;
var metersToMiles = 0.000621371;
var feetToKilometers = 0.0003048;
var metersToKilometers = 1000;
var metersToFeet = 3.28084;

module.exports = function(opts) {
    /**
    * Configuration options
    * @param {object} `units` - either `miles` or `km`. `maxReRouteDistance` - max distance the user can be from the route. `maxSnapToLocation` - max distance to snap user to route. `completionDistance` - distance away from end of step that is considered a completion. If this distance is shorter than the step distance, it will be changed to 10ft. `warnUserTime` - number of seconds ahead of maneuver to warn user about maneuver. `shortCompletionDistance` - if the step is shorter than the `completionDistance`, this distance will be used to calculate if the step has been completed. `userBearingCompleteThreshold` - Bearing threshold for the user to complete a step
    */
    var options = {
        units: opts.units || 'miles',
        warnUserTime: opts.warnUserTime || 30, // seconds
        userBearingCompleteThreshold: opts.userBearingCompleteThreshold || 30
    };

    options.maxReRouteDistance = opts.maxReRouteDistance ? opts.maxReRouteDistance : opts.units === 'miles' ? 150 * feetToMiles : 150 * feetToKilometers;
    options.maxSnapToLocation = opts.maxSnapToLocation ? opts.maxSnapToLocation : opts.units === 'miles' ? 50 * feetToMiles : 50 * feetToKilometers;
    options.completionDistance = opts.completionDistance ? opts.completionDistance : opts.units === 'miles' ? 50 * feetToMiles : 50 * feetToKilometers;
    options.shortCompletionDistance = opts.shortCompletionDistance ? opts.shortCompletionDistance : opts.units === 'miles' ? 10 * feetToMiles : 10 * feetToKilometers;

    /**
     * Given a user location and route, calculates closest step to user.
     * @param {object} user point feature representing user location. Must be a valid GeoJSON object.
     * @param {object} route from [Mapbox directions API](https://www.mapbox.com/developers/api/directions/).
     * The Mapbox directions API returns an object with up to 2 `routes` on the `route` key. `findNextStep` expects of these routes, either the first or second.
     * @param {number} userCurrentStep along the route
     * @param {number} userBearing current user bearing. If provided, the user must be within a certain threadhold of the steps exit bearing to successful complete a step.
     * @returns {object} Containing 3 keys: `step`, `distance`, `snapToLocation`. `distance` is the line distance to end of step, `absoluteDistance` is the users absolute distance to the end of the route `snapToLocation` is location along route which is closest to the user.
     */
    function findNextStep(user, route, userCurrentStep, userBearing) {
        var currentStep = {};
        var stepCoordinates = route.steps[userCurrentStep].geometry.coordinates;
        var segmentRoute = {
            type: 'Feature',
            geometry: {
                type: 'LineString',
                coordinates: stepCoordinates
            }
        };

        var closestPoint = turfPointOnLine(segmentRoute, user);
        var distance = turfDistance(user, closestPoint, options.units);

        var segmentEndPoint = { type: 'Feature', geometry: { type: 'Point', coordinates: stepCoordinates[stepCoordinates.length - 1] }};
        var segmentSlicedToUser = turfLineSlice(user, segmentEndPoint, segmentRoute);
        var userDistanceToEndStep = turfLineDistance(segmentSlicedToUser, options.units);

        //
        // Check if user has completed step. Two factors:
        //   1. Are they within a certain threshold of the end of the step?
        //   2. If a bearing is provided, is their bearing within a current threshold of the exit bearing for the step
        //
        var stepDistance = options.units === 'miles' ? route.steps[userCurrentStep].distance * metersToMiles : stepCoordinates[userCurrentStep].distance * metersToKilometers;
        // If the step distance is less than options.completionDistance, modify it and make it 10 ft
        var modeifiedCompletionDistance = stepDistance < options.completionDistance ? options.shortCompletionDistance : options.completionDistance;
        // Check if users bearing is within threshold of the steps exit bearing
        var withinBearingThreshold = userBearing ? Math.abs(userBearing - route.steps[userCurrentStep].maneuver.bearing_after) <= options.userBearingCompleteThreshold ? true : false : true;
        currentStep.step = withinBearingThreshold && (userDistanceToEndStep < modeifiedCompletionDistance) ? userCurrentStep + 1 : userCurrentStep; // Don't set next step + 1 if at the end of the route

        currentStep.distance = userDistanceToEndStep;
        currentStep.shouldReRoute = turfDistance(user, closestPoint, options.units) > options.maxReRouteDistance ? true : false;
        currentStep.stepDistance = stepDistance;
        currentStep.absoluteDistance = turfDistance(user, segmentEndPoint, options.units);
        currentStep.snapToLocation = distance < opts.maxSnapToLocation ? closestPoint : user;

        // Alert levels
        currentStep.alertUserLevel = {
            low: userDistanceToEndStep < 1 && route.steps[userCurrentStep].distance * metersToMiles > 1, // Step must be longer than 1 miles
            high: (userDistanceToEndStep < 150 * feetToMiles) && route.steps[userCurrentStep].distance * metersToFeet > 150 // Step must be longer than 150 ft
        };
        return currentStep;
    };

    return {
        findNextStep: findNextStep
    };
};


lib/current-step.js
var point = require('turf-point');
var pointOnLine = require('turf-point-on-line');

module.exports = currentStep;
function currentStep(user, route) {
  var location = point(user.geometry.coordinates);
  var distances = [];
  var steps = route.routes[0].legs[0].steps;
  for (var i = 0; i < steps.length; i++) {
    var feature = '{"type":"Feature","properties":{},"geometry":' + JSON.stringify(steps[i].geometry) + '}';
    var distance = pointOnLine(JSON.parse(feature),location).properties.dist;
    distances.push(distance);
  }
  return indexOfMin(distances);

  function indexOfMin(array) {
    if (array.length === 0) { return -1; }
    var min = array[0];
    var minIndex = 0;
    for (var i = 1; i < array.length; i++) {
      if (array[i] < min) {
        minIndex = i;
        min = array[i];
      }
    }
    return minIndex;
  }
}

index.js
var turfPointOnLine = require('turf-point-on-line');
var turfDistance = require('turf-distance');
var turfLineDistance = require('turf-line-distance');
var turfLineSlice = require('turf-line-slice');
var polyline = require('polyline');

var feetToMiles = 0.000189394;
var metersToMiles = 0.000621371;
var feetToKilometers = 0.0003048;
var metersToKilometers = 1000;
var metersToFeet = 3.28084;

module.exports = function(opts) {
    /**
    * Configuration options
    * @param {object} `units` - either `miles` or `km`. `maxReRouteDistance` - max distance the user can be from the route. `maxSnapToLocation` - max distance to snap user to route. `completionDistance` - distance away from end of step that is considered a completion. If this distance is shorter than the step distance, it will be changed to 10ft. `warnUserTime` - number of seconds ahead of maneuver to warn user about maneuver. `shortCompletionDistance` - if the step is shorter than the `completionDistance`, this distance will be used to calculate if the step has been completed. `userBearingCompleteThreshold` - Bearing threshold for the user to complete a step
    */
    var options = {
        units: opts.units || 'miles',
        warnUserTime: opts.warnUserTime || 30, // seconds
        userBearingCompleteThreshold: opts.userBearingCompleteThreshold || 30
    };

    options.maxReRouteDistance = opts.maxReRouteDistance ? opts.maxReRouteDistance : opts.units === 'miles' ? 150 * feetToMiles : 150 * feetToKilometers;
    options.maxSnapToLocation = opts.maxSnapToLocation ? opts.maxSnapToLocation : opts.units === 'miles' ? 50 * feetToMiles : 50 * feetToKilometers;
    options.completionDistance = opts.completionDistance ? opts.completionDistance : opts.units === 'miles' ? 50 * feetToMiles : 50 * feetToKilometers;
    options.shortCompletionDistance = opts.shortCompletionDistance ? opts.shortCompletionDistance : opts.units === 'miles' ? 10 * feetToMiles : 10 * feetToKilometers;

    /**
     * Given a user location and route, calculates closest step to user.
     * @param {object} user point feature representing user location. Must be a valid GeoJSON object.
     * @param {object} route from [Mapbox directions API](https://www.mapbox.com/developers/api/directions/).
     * The Mapbox directions API returns an object with up to 2 `routes` on the `route` key. `getCurrentStep` expects of these routes, either the first or second.
     * @param {number} userCurrentStep along the route
     * @param {number} userBearing current user bearing. If provided, the user must be within a certain threadhold of the steps exit bearing to successful complete a step.
     * @returns {object} Containing 3 keys: `step`, `distance`, `snapToLocation`. `distance` is the line distance to end of step, `absoluteDistance` is the users absolute distance to the end of the route `snapToLocation` is location along route which is closest to the user.
     */
    function getCurrentStep(user, route, userCurrentStep, userBearing) {
        var currentStep = {};
        var stepCoordinates;
        if (typeof route.steps[userCurrentStep].geometry === 'string') {
            stepCoordinates = polyline
                .decode(route.steps[userCurrentStep].geometry)
                .map(function(coordinate) {
                    return [coordinate[1], coordinate[0]];
                });
        } else {
            stepCoordinates = route.steps[userCurrentStep].geometry.coordinates;
        }
        var segmentRoute = {
            type: 'Feature',
            geometry: {
                type: 'LineString',
                coordinates: stepCoordinates
            }
        };

        var closestPoint = turfPointOnLine(segmentRoute, user);
        var distance = turfDistance(user, closestPoint, options.units);

        var segmentEndPoint = { type: 'Feature', geometry: { type: 'Point', coordinates: stepCoordinates[stepCoordinates.length - 1] }};
        var segmentSlicedToUser = turfLineSlice(user, segmentEndPoint, segmentRoute);
        var userDistanceToEndStep = turfLineDistance(segmentSlicedToUser, options.units);

        //
        // Check if user has completed step. Two factors:
        //   1. Are they within a certain threshold of the end of the step?
        //   2. If a bearing is provided, is their bearing within a current threshold of the exit bearing for the step
        //
        var stepDistance = options.units === 'miles' ? route.steps[userCurrentStep].distance * metersToMiles : stepCoordinates[userCurrentStep].distance * metersToKilometers;
        // If the step distance is less than options.completionDistance, modify it and make it 10 ft
        var modifiedCompletionDistance = stepDistance < options.completionDistance ? options.shortCompletionDistance : options.completionDistance;
        // Check if users bearing is within threshold of the steps exit bearing
        var withinBearingThreshold = userBearing ? Math.abs(userBearing - route.steps[userCurrentStep].maneuver.bearing_after) <= options.userBearingCompleteThreshold ? true : false : true;

        // Do not increment userCurrentStep if the user is approaching the final step
        if (userCurrentStep < route.steps.length - 2) {
            currentStep.step = withinBearingThreshold && (userDistanceToEndStep < modifiedCompletionDistance) ? userCurrentStep + 1 : userCurrentStep; // Don't set next step + 1 if at the end of the route
        } else {
            currentStep.step = userCurrentStep;
        }

        currentStep.distance = userDistanceToEndStep;
        currentStep.stepDistance = stepDistance;
        currentStep.shouldReRoute = turfDistance(user, closestPoint, options.units) > options.maxReRouteDistance ? true : false;
        currentStep.absoluteDistance = turfDistance(user, segmentEndPoint, options.units);
        currentStep.snapToLocation = distance < opts.maxSnapToLocation ? closestPoint : user;

        // Alert levels
        currentStep.alertUserLevel = {
            low: userDistanceToEndStep < 1 && route.steps[userCurrentStep].distance * metersToMiles > 1, // Step must be longer than 1 miles
            high: (userDistanceToEndStep < 150 * feetToMiles) && route.steps[userCurrentStep].distance * metersToFeet > 150 // Step must be longer than 150 ft
        };
        return currentStep;
    };

    return {
        getCurrentStep: getCurrentStep
    };
};
